src/controller.rs

```diff
@@ -335,10 +457,107 @@ async fn launch_job(
 
     handle_resource_exists(jobs.create(&pp, &job).await)?;
 
-    // TODO:
+    Ok(())
+}
+
+// kube crate comes with is_job_completed but that condition is true only if it completes successfully.
+fn is_job_failed() -> impl Condition<Job> {
+    |obj: Option<&Job>| {
+        if let Some(job) = &obj {
+            if let Some(s) = &job.status {
+                if let Some(conds) = &s.conditions {
+                    if let Some(pcond) = conds.iter().find(|c| c.type_ == "Failed") {
+                        return pcond.status == "True";
+                    }
+                }
+            }
+        }
+        false
+    }
+}
+
+async fn capture_logs(app_instance: &AppInstance, ctx: &Context, job_uid: String) -> Result<()> {
+    let ns = &app_instance.namespace().ok_or(Error::NamespaceRequired)?;
+    info!(?ns, "reporting errors");
+
+    let pods_api: Api<Pod> = Api::namespaced(ctx.client.clone(), ns);
+    let job_name = job_name_for(app_instance);
+
+    let pods = pods_api
+        .list(&ListParams {
+            label_selector: Some(format!("job-name={job_name},controller-uid={job_uid}")),
+            ..Default::default()
+        })
+        .await?;
+
+    let mut per_container_logs = HashMap::new();
+
+    // There should be exactly one pod per job. In the unlikely even
+    // something is broken with k8s and we end up getting two pods matching the same job uid
+    // let's just get the logs of all these pods and concatenate them together. Chances are
+    // that this is easier to debug than just getting logs for a random pod.
     //
-    // 1. if job exists check if it's has terminated and take action
-    // 2. make sure we watch the job as it changes status
+    // NOTE(mkm): I don't know how likely this is to happen so I'm not sure if it's worth doing
+    // something more complicated like capturing the pod names and grouping the logs by pod name.
```

Comment:
Agreed with this.

I _think_ we're covered by the `spec.completions` and `spec.parallelism` both being unset, which leads to a single `Job` running and no duplicates, but in the event it does occur I think this is useful for troubleshooting 👍

Reply:
I noticed that initially we had backoffLimit set to 1, which intuitively seems to mean that only one pod can exist, but in reality meant that if a pod fails, it will retry once, so when the pod fails there will be left two pods sitting around by the time the job fails.

now, we fixed this misconfig in particular, but I was thinking: what happens if the job controller tries to create a pod that never starts because it lands on a node that is misbehaving? will it honour the backoff or will it end up having two pods anyway?